dipy/core/optimize.py

```diff
@@ -342,3 +347,181 @@ def fit(self, X, y):
+        self.L = L
+
+        # Problem size
+        t = len(A) if A else 0
+        k = t - m - 1
+
+        sparsity = [(i, j) for i in range(m) for j in range(i, m)]
+
+        # Unknowns
+        self._X = cvxpy.Parameter((m, m), sparsity=sparsity)    # Design matrix
+        self._f = cvxpy.Parameter(m)    # Given solution for feasibility check
+        self._h = cvxpy.Variable(m)     # Solution to constrained problem
+        self._y = cvxpy.Parameter(m)    # Regressand
+
+        # Error output
+        self._zeros = np.zeros(m)
+
+        # Objective
+        if Version(cvxpy.__version__) < Version('1.1'):
+            c = self._X*self._h - self._y
+            if L is not None:
+                c += L*self._h
+        else:
+            c = self._X@self._h - self._y
+            if L is not None:
+                c += L@self._h
+
+        f_objective = cvxpy.Minimize(0)
+        p_objective = cvxpy.Minimize(cvxpy.norm(c))
+
+        # Constraints
+        if t:
+            M = F = A[0]
+            if k > 0:
+                for i in range(m):
+                    F += self._f[i] * A[i + 1]
+                    M += self._h[i] * A[i + 1]
+                self._s = cvxpy.Variable(k)
+                for j in range(k):
+                    F += self._s[j] * A[m + j + 1]
+                    M += self._s[j] * A[m + j + 1]
+            else:
+                for i in range(t - 1):
+                    F += self._f[i] * A[i + 1]
+                    M += self._h[i] * A[i + 1]
+            f_constraints = [F >> 0]
+            p_constraints = [M >> 0]
+        else:
+            f_constraints = p_constraints = []
+
+        # CVXPY problems
+        self.problem = cvxpy.Problem(p_objective, p_constraints)
+        self.unconstrained_problem = cvxpy.Problem(p_objective)
+        self.feasibility_problem = cvxpy.Problem(f_objective, f_constraints)
+
+    def solve(self, design_matrix, measurements, check=False, **kwargs):
+        r""" Solve CVXPY problem
+
+        Solve a CVXPY problem instance for a given design matrix and a given set
+        of observations, and return the optimum.
+
+        Parameters
+        ----------
+        design_matrix : array (n, m)
+            Design matrix.
+        measurements : array (n)
+            Measurements.
+        check : boolean (optional)
+            If True check whether the unconstrained optimization solution
+            already satisfies the constraints, before running the constrained
+            optimization. This adds overhead, but can avoid unnecessary
+            constrained optimization calls.
+            Default: False
+        kwargs : keyword arguments
+            Arguments passed to the CVXPY solve method.
+
+        Returns
+        -------
+        h : array (m)
+             Estimated optimum for problem variables $h$.
+        """
+
+        # Compute and set reduced problem parameters
+        X = np.linalg.cholesky(np.dot(design_matrix.T, design_matrix)).T
```

Comment:
Hello, 

Had a look at the code and seems good to me :) 

I just have a suggestion for this line: doing the cholesky decomposition requires the matrix to be strictly positive definite. While this shouldn't be a problem for MAP, other methods (especially those requiring tensor encoded data) might get a rank deficient matrix which is not directly decomposable with cholesky. If the plan is to reuse this code for several other methods, maybe using the more general way of computing the matrix square root would be better. Or in general, a warning should be issued to inform the user that the system is underdetermined and the solution is not unique.

Reply:
Nice catch. I've added a warning in the case of a rank deficient design matrix. I'm not sure if there is a single satisfactory way to handle rank deficiency for other models, so for now I think just raising a warning to the user is the way to go.